cirq/google/calibration/workflow.py

```diff
@@ -11,21 +11,210 @@
+                    f'and {translated_gate}'
+                )
+            else:
+                gate = translated_gate
+            pair = cast(
+                Tuple[Qid, Qid], tuple(sorted(op.qubits) if pairs_in_canonical_order else op.qubits)
+            )
+            pairs.append(pair)
+
+    if gate is None:
+        # Either empty, single-qubit or measurement moment.
+        return None
+
+    if gate is not None and (measurement or single_qubit):
+        raise IncompatibleMomentError(
+            f'Moment contains mixed two-qubit operations and '
+            f'single-qubit operations or measurement operations.'
+        )
+
+    return FloquetPhasedFSimCalibrationRequest(
+        pairs=tuple(sorted(pairs) if pairs_sorted else pairs), gate=gate, options=options
+    )
+
+
+def floquet_characterization_for_circuit(
+    circuit: Circuit,
+    options: FloquetPhasedFSimCalibrationOptions = FloquetPhasedFSimCalibrationOptions.all_except_for_chi_options(),
+    gates_translator: Callable[[Gate], Optional[FSimGate]] = sqrt_iswap_gates_translator,
+    merge_sub_sets: bool = True,
+    initial: Optional[Tuple[List[FloquetPhasedFSimCalibrationRequest], List[Optional[int]]]] = None,
+) -> Tuple[List[FloquetPhasedFSimCalibrationRequest], List[Optional[int]]]:
+    """Extracts a minimal set of Floquet characterization requests necessary to characterize given
+    circuit.
+
+    The circuit can only be composed of single qubit operations, measurement operations and
+    operations supported by gates_translator.
+
+    Args:
+        circuit: Circuit to characterize.
+        options: Options that are applied to each characterized gate within a moment. Defaults
+            to all_except_for_chi_options which is the broadest currently supported choice.
+        gates_translator: Function that translates a gate to a supported FSimGate which will undergo
+            characterization. Defaults to sqrt_iswap_gates_translator.
+        merge_sub_sets: Whether to merge moments that can be characterized at the same time
+            together.
+        initial: The characterization requests obtained by a previous scan of another circuit. This
+            might be used to find a minimal set of moments to characterize across many circuits.
+
+    Returns:
+        Tuple of:
+          - list of calibration requests,
+          - list of indices of the generated characterization requests for each moment in the
+            supplied circuit. If None occurs at certain position, it means that the related moment
+            does not require characterization.
+
+    Raises:
+        IncompatibleMomentError when circuit contains a moment with operations other than the
+        operations matched by gates_translator, or it mixes a single qubit and two qubit gates.
+    """
+
+    def append_if_missing(calibration: FloquetPhasedFSimCalibrationRequest) -> int:
+        if calibration.pairs not in pairs_map:
```

Comment:
`pairs_map` undefined?

Reply:
I guess this piece of code is not really nice and I should refactor it... I also need to spend some time with explaining what's going on.

Reply:
Done.